app/src/main/java/com/github/orelzion/spotifyclone/view/TracksListFragment.kt

```diff
@@ -0,0 +1,126 @@
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.activityViewModels
+import androidx.fragment.app.commit
+//import androidx.fragment.app.viewModels
+import androidx.recyclerview.widget.RecyclerView
+import com.github.orelzion.spotifyclone.R
+import com.github.orelzion.spotifyclone.viewmodel.TrackViewData
+//import com.github.orelzion.spotifyclone.model.Tracks
+//import com.github.orelzion.spotifyclone.model.TracksResponse
+//import com.github.orelzion.spotifyclone.model.repository.BrowseRepository
+//import com.github.orelzion.spotifyclone.viewmodel.TrackViewData
+import com.github.orelzion.spotifyclone.viewmodel.TracksViewModel
+
+
+/**
+ * A fragment representing a list of Tracks.
+ * Fragment is at the View level
+ */
+class TracksListFragment : Fragment(R.layout.fragment_items_list) {
+    /**
+     * View level
+     */
+
+    /**
+     * Reference to RecyclerView that holds tracks' list
+     */
+    private val tracksListView: RecyclerView by lazy { requireView().findViewById(R.id.itemsList) }
+
+    /**
+     * Reference to an adapter for the RecyclerView
+     */
+    private val tracksAdapter: TracksAdapter by lazy { TracksAdapter() }
+
+
+    /**
+     * ViewModel level
+     */
+
+    /**
+     * Reference to corresponding ViewModel
+     * We use activityViewModels rather than viewModels because we don't want attach that ViewModel
+     * to the current Fragment.
+     */
+    private val tracksViewModel by activityViewModels<TracksViewModel>()
+
+    /**
+     * companion object is an object created once for a class; it is not being duplicated for each instance
+     */
+    companion object {
+        /**
+         * factory method
+         * accessed via TracksListFragment, as it is being declared in companion object
+         */
+        fun newInstance(albumId: String): TracksListFragment {
+            return TracksListFragment().apply {
+                arguments = Bundle().apply {
+                    putString("albumId", albumId)
+                }
+            }
+        }
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        tracksListView.adapter = tracksAdapter.apply {
+            clickListener = this@TracksListFragment::onItemClicked
+        }
+
+        /**
+         * From Bundle (created in AlbumsListFragment, when an album was clicked),
+         * get album's id ->
+         */
+        val albumId = requireArguments().getString("albumId")
+        /**
+         * -> and load it's tracks (loadTracks() would call ViewModel's loadTracks())
+         */
+        if (savedInstanceState == null) {
+            albumId?.let {
+                loadTracks(it)
+            }
+        }
+        // TODO: add else - restore savedInstanceState rather than (re-)loadSongs
+
+        // hide progressbar
+        requireView().findViewById<ProgressBar>(R.id.progressBar).isVisible = false
```

Comment:
same as before, maybe it's better to hide **after** you received the tracks

Reply:
I assume that placing the hiding loadTracks() does not guarantee their receiving; right?
Then I'm not sure I know what to check in order to make sure that tracks(/albums) are received already.

Reply:
Inside our observer, we know for sure that the request is done, so we hide the progress bar